### DISCUSSION TITLE
using commit event with async environment

### DISCUSSION THREAD

**Discussion by adsko** [sqlalchemy/sqlalchemy#6527] [Usage Questions] [answered]
Hello,

I have a question about events and async connectors. I want to call my own pub/sub service only when a transaction is commited correctly. I'm using AsyncEngine, and flow looks like:

```py
async with db.begin() as conn:
    await put_to_db(conn)


async def put_to_db(conn: AsyncConnection):
    await conn.execute(/*code*/)
    pub_sub_service.inform(conn)

# pub sub service
async def inform(conn):
    if conn.in_transaction():
        event.listen(conn.sync_connection, "commit", _on_commit)
        event.listen(conn.sync_connection, "rollback", _on_rollback)
    else:
         _inform_now()

def _on_commit(*args, **kwargs):
    # Code for commit

def _on_rollback(*args, **kwargs):
   # Code for rollback
```
(This code shows only flow of an app)

And there I've some questions:
 * Are these events called only for current transaction, or can be called for different transaction if second request is handled in the same time (ASGI)?
 * Is there an event like after_commit for the connection? As I read in a code, currently the commit is called before db commit, so the inform can be called before real data are saved
 * Is it safer and better to use AsyncSession in situation like this, even if I'm not using ORM?

**Comment by zzzeek** [comment, marked as answer]
> Hello,
> 
> I have a question about events and async connectors. I want to call my own pub/sub service only when a transaction is commited correctly. I'm using AsyncEngine, and flow looks like:
> 
> ```python
> async with db.begin() as conn:
>     await put_to_db(conn)
> 
> 
> async def put_to_db(conn: AsyncConnection):
>     await conn.execute(/*code*/)
>     pub_sub_service.inform(conn)
> 
> # pub sub service
> async def inform(conn):
>     if conn.in_transaction():
>         event.listen(conn.sync_connection, "commit", _on_commit)
>         event.listen(conn.sync_connection, "rollback", _on_rollback)
>     else:
>          _inform_now()
> 
> def _on_commit(*args, **kwargs):
>     # Code for commit
> 
> def _on_rollback(*args, **kwargs):
>    # Code for rollback
> ```
> 
> (This code shows only flow of an app)
> 
> And there I've some questions:
> 
>     * Are these events called only for current transaction, or can be called for different transaction if second request is handled in the same time (ASGI)?

assuming you are using the "with engine.begin() as conn" pattern, the "conn" here is the Connection that is produced on-demand within the engine.begin() method.   When the context manager ends, the transaction is committed or rolled back, and then this Connection object is disposed; the underlying DBAPI connection is the part that goes back to a connection pool.   Therefore, if you apply an event listener to that specific "conn" object only, when the "engine.begin()" block ends, that Connection is disposed and so is the event listener which would allow "listener per transaction".

If OTOH you called "engine.connect()" and then used "commit as you go" on the Connection, calling .rollback() and .commit() in an ongoing way against that same Connection object, that event listener would stay active for each of those events.


> 
>     * Is there an event like after_commit for the connection? As I read in a code, currently the commit is called before db commit, so the inform can be called before real data are saved

there's not as the purpose of events is to allow user code to invoke at stages that are otherwise hard to get towards.  if you want to do something after the commit, just put it in your code:

```python
with engine.begin() as conn:
    # do things
do_my_after_commit()
```

(edit) if you're looking to use this pattern repeatedly you can make your own context manager (would need to use the async version of contextlib etc., which is available, but this is the general idea):

```python
import contextlib

@contextlib.contextmanagr
def do_my_thing(engine):
    with engine.begin() as conn:
        yield conn
    do_my_after_commit()
```

then code would look like:

```python
with do_my_thing() as conn:
   # stuff
```



> 
>     * Is it safer and better to use AsyncSession in situation like this, even if I'm not using ORM?

People seem to find the Session interface more familiar, even though it is more complicated.   it does have an after_commit() method so that might be useful here.  I dont think there's a single answer to that question though.

**Reply by adsko** [reply]
As you have mentioned, it's better to extract send information to commit phase in the code. 

I think, I'll go with context like:
```py
async with pub_sub.session() as ps_session:
  async with db.begin() as async_connection:
     do_db_call(async_connection, ps_session)
```

It's nice that event is called for specific connection only, but makes issues, because commit is not really commit but it'll be if everything goes ok. Because of it, I need to check in a loop if connection is still during commit.

I'm not sure, what'll hapen, when db commit throws error, but as long as I will not use it, it's fine :) 

Thanks for your replay.
